File: tests/node/benchmark.js
const NoTorToday = require('../../core/node/NoTorToday');
const fs = require('fs');
const path = require('path');

async function benchmark() {
  const iterations = 1000;
  
  // Initialize
  const checker = new NoTorToday(false);
  
  // Load sample data for offline testing
  try {
    const sampleData = fs.readFileSync(
      path.join(__dirname, '../../datasets/sample-exit-addresses.txt'),
      'utf8'
    );
    checker.parseExitAddresses(sampleData);
  } catch (e) {
    // Try to download if sample not available
    try {
      await checker.updateList();
    } catch (err) {
      console.error('Cannot load data:', err.message);
      process.exit(1);
    }
  }
  
  const testIPs = [
    '185.220.101.1',
    '8.8.8.8',
    '1.1.1.1',
    '192.168.1.1',
    '10.0.0.1',
  ];
  
  // Warmup
  for (let i = 0; i < 100; i++) {
    testIPs.forEach(ip => checker.isTor(ip));
  }
  
  // Benchmark
  const start = process.hrtime.bigint();
  
  for (let i = 0; i < iterations; i++) {
    testIPs.forEach(ip => checker.isTor(ip));
  }
  
  const end = process.hrtime.bigint();
  const durationMs = Number(end - start) / 1_000_000;
  const opsPerSec = (iterations * testIPs.length) / (durationMs / 1000);
  
  console.log(JSON.stringify({
    language: 'Node.js',
    duration_ms: durationMs.toFixed(2),
    operations: iterations * testIPs.length,
    ops_per_second: Math.round(opsPerSec),
    exit_nodes_loaded: checker.getCount()
  }));
}

benchmark().catch(console.error);

File: tests/go/benchmark.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"

	"github.com/TheusHen/NoTorToday/core/go"
)

func main() {
	iterations := 1000

	// Initialize
	checker, err := notortoday.New(false, "")
	if err != nil {
		panic(err)
	}

	// Load sample data for offline testing
	samplePath := filepath.Join("../../datasets/sample-exit-addresses.txt")
	data, err := ioutil.ReadFile(samplePath)
	if err != nil {
		// Try to update if sample not available
		if err := checker.UpdateList(); err != nil {
			fmt.Fprintf(os.Stderr, "Cannot load data: %v\n", err)
			os.Exit(1)
		}
	} else {
		checker.ParseExitAddresses(string(data))
	}

	testIPs := []string{
		"185.220.101.1",
		"8.8.8.8",
		"1.1.1.1",
		"192.168.1.1",
		"10.0.0.1",
	}

	// Warmup
	for i := 0; i < 100; i++ {
		for _, ip := range testIPs {
			checker.IsTor(ip)
		}
	}

	// Benchmark
	start := time.Now()

	for i := 0; i < iterations; i++ {
		for _, ip := range testIPs {
			checker.IsTor(ip)
		}
	}

	duration := time.Since(start)
	durationMs := float64(duration.Nanoseconds()) / 1000000.0
	opsPerSec := int(float64(iterations*len(testIPs)) / (durationMs / 1000.0))
	
	// Handle edge case where duration is too small
	if durationMs < 0.01 {
		durationMs = 0.01
		opsPerSec = iterations * len(testIPs) * 100
	}

	result := map[string]interface{}{
		"language":           "Go",
		"duration_ms":        fmt.Sprintf("%.2f", durationMs),
		"operations":         iterations * len(testIPs),
		"ops_per_second":     opsPerSec,
		"exit_nodes_loaded":  checker.GetCount(),
	}

	jsonData, _ := json.Marshal(result)
	fmt.Println(string(jsonData))
}
